### ISSUE TITLE
Allow history to be used with masked edit controls

### ISSUE THREAD

**Issue by rohitab** [FarGroup/FarManager#617]
### Description of the new feature or improvement

If you create a fixed edit control `DI_FIXEDIT` and enable mask `DIF_MASKEDIT` and history `DIF_HISTORY`, then the mask does not work.

This appears to be due to the following code, which removes the mask if history is enabled:

https://github.com/FarGroup/FarManager/blob/a2ba494e12a78423141db73eef166af0077c18da/far/dialog.cpp#L862-L866

The translation of the above comment is: `// DIF_HISTORY has a higher precedence than DIF_MASKEDIT`

I'm not sure why this restriction was put in place. For testing, I removed lines `865` and `886` from the above code and rebuilt _Far_. The masked control now works with history enabled. You can press `Ctrl+Up` or `Ctrl+Down` to open the history dropdown. The only issue is that auto-complete does not work.

Reference: [#193 Hexitor - Enable history for dialog inputs](https://sourceforge.net/p/farplugs/tickets/193/)

**Comment by alabuzhev** [comment]
A long time ago in a Far far, far away there was a little struct to describe a dialog item:
https://github.com/FarGroup/FarManager/blob/3f560a1d5f5b25f0795f92fef4d927fe839e2223/far/plugin.hpp#L118-L127
There were no histories, there were no fixed edits, life was nice and simple.

Some time later, someone came up with the idea of ​​histories.
But how to retrofit a history name into the existing struct without breaking the ABI?
FarDialogItem was not designed to be extensible, it does not have `int StructSize;` or anything like that.
So what did they do? This:
https://github.com/FarGroup/FarManager/blob/13565fd69575a673df40431a03ddb03ec5390e7a/far/plugin.hpp#L328-L341
They repurposed an existing field that was not used for edit fields - `int Selected;`, turned it into a union, allowing to specify the selection state for buttons and the history name for edit controls, yay!

3 weeks later, someone came up with the idea of ​​masked edits. Where to put the mask? Into the same union of course!
https://github.com/FarGroup/FarManager/blob/95ff278689994bca30be49b535da79205c7cac31/far/plugin.hpp#L380-L394

But, obviously, using the same address for both `char *History;` and `char *Mask;` means that you cannot have both. There was a short discussion about it, and the outcome was that looking for alternative solutions is not worth the effort, because "no one needs histories there anyway".

And this is how the current logic was born.

------

Fast forward from 2000 to 2011.
Far 2 has been released and Far 3 has just begun.
We have one in a once-in-a-lifetime chance to break the world and correct some mistakes of the past.
And this is exactly what happened:
https://github.com/FarGroup/FarManager/blob/1386cfff0724d05bf03e5caa312117cc998e6321/far/plugin.hpp#L459-L482

History and Mask are independent entities now and nothing stops us from having the cake and eating it too. Awesome.

"Wait, but if it was deliberately allowed over 10 years ago, why it still does not work?" you might ask.
Because of that little piece of code from 2000 that clears the flag.
Why is it still there? I don't know, but I think the intention of the change in 3.0.1836 was pretty clear - "it should be possible", so quite probably just an overlook.

To summarize: I can't see why not, especially since you've already tested it.
Would you like to submit a PR? (for attribution's sake 😄)  If no, I can do it myself.

**Comment by rohitab** [comment]
@alabuzhev Thanks for the detailed explanation. Your posts are always fun to read!

I appreciate you giving me the opportunity to contribute here. I've submitted a PR. It's my first one, so apologies in advance for any issues.

**Comment by alabuzhev** [comment]
Thanks, merged.

**Comment by rohitab** [comment]
I just tested version `3.0.6098.0 x64`. The issue is fixed. Thank you.
